ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/ValidateDisclosedContracts.scala

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.validation
+
+import com.daml.api.util.TimestampConversion
+import com.daml.error.ContextualizedErrorLogger
+import com.daml.error.definitions.LedgerApiErrors
+import com.daml.lf.command.{ContractMetadata, DisclosedContract}
+import com.daml.lf.data.ImmArray
+import com.daml.ledger.api.v1.commands.{Commands => ProtoCommands}
+import com.daml.ledger.api.v1.commands.{DisclosedContract => ProtoDisclosedContract}
+import com.daml.ledger.api.validation.ValidationErrors.invalidArgument
+import com.daml.ledger.api.validation.ValueValidator.{
+  validateOptionalIdentifier,
+  validateRecordFields,
+}
+import com.daml.lf.value.Value.ValueRecord
+import com.daml.platform.server.api.validation.FieldValidations.{
+  requireContractId,
+  requirePresence,
+  validateHash,
+  validateIdentifier,
+}
+import com.google.protobuf.timestamp.Timestamp
+import io.grpc.StatusRuntimeException
+import com.daml.lf.data.Time
+
+import scala.collection.mutable
+import scala.util.Try
+
+class ValidateDisclosedContracts(featureEnabled: Boolean) {
+  def apply(commands: ProtoCommands)(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): Either[StatusRuntimeException, ImmArray[DisclosedContract]] =
+    for {
+      _ <- Either.cond(
+        featureEnabled || commands.disclosedContracts.isEmpty,
+        (),
+        LedgerApiErrors.RequestValidation.InvalidField
+          .Reject(
+            "disclosed_contracts",
+            "feature in development: disclosed_contracts should not be set",
+          )
+          .asGrpcError,
+      )
+      validatedDisclosedContracts <- validateDisclosedContracts(commands.disclosedContracts)
+    } yield validatedDisclosedContracts
+
+  private def validateDisclosedContracts(
+      disclosedContracts: Seq[ProtoDisclosedContract]
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): Either[StatusRuntimeException, ImmArray[DisclosedContract]] = {
+    type ZeroType =
+      Either[
+        StatusRuntimeException,
+        mutable.Builder[DisclosedContract, ImmArray[DisclosedContract]],
+      ]
+
+    disclosedContracts
+      .foldLeft[ZeroType](Right(ImmArray.newBuilder))((contractz, contract) =>
```

Comment:
Out of interest, any specific reason to user `ImmArray.newBuilder` rather than just building up a List or Seq?